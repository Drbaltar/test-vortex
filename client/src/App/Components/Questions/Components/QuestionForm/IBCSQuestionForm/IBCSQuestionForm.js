import React from 'react';

import BodyCard from '../../../../shared-components/BodyCard/BodyCard';
import BaseQuestionFields from '../BaseQuestionFields/BaseQuestionFields';
import ErrorMessage from '../../../../shared-components/ErrorMessage';

class IBCSQuestionForm extends React.Component {
    constructor (props) {
        super(props);

        this.FormFooter = this.props.children;

        this.initialState = this.getBlankInitialState();
        this.state = this.initialState;
    }

    getBlankInitialState = () => {
        return {
            questionType: 'Multiple Choice',
            questionDescription: '',
            correctAnswer: '',
            answerA: '',
            answerB: '',
            answerC: '',
            topic: {
                majorCategory: '',
                subCategory: ''
            },
            inputValidity: this.getBlankInputValidity(),
            isTopicLoading: false,
            existingTopicCategories: null
        };
    }

    getBlankInputValidity = () => {
        return {
            isQuestionTypeValid: null,
            isQuestionDescriptionValid: null,
            isCorrectAnswerValid: null,
            isAnswerAValid: null,
            isAnswerBValid: null,
            isAnswerCValid: null,
            isTestTypeValid: null,
            isMajorTopicValid: null,
            isSubTopicValid: null
        };
    }

    render() {
        return (
            <BodyCard title={this.props.title}>
                {this.getFormBody()}
                {this.getFormFooter()}
            </BodyCard>
        );
    }

    getFormBody = () => {
        return (
            <div id='issue-form-body'>
                {this.getBaseQuestionFields()}
                {this.props.errorMessage ? this.getErrorMessage() : null}
            </div>
        );
    }

    getBaseQuestionFields = () => {
        return <BaseQuestionFields inputHandler={(e) => this.handleInputChange(e)}
            questionType={this.state.questionType} isQuestionTypeValid={this.state.inputValidity.isQuestionTypeValid}
            questionDescription={this.state.questionDescription} isQuestionDescriptionValid={this.state.inputValidity.isQuestionDescriptionValid}
            correctAnswer={this.state.correctAnswer} isCorrectAnswerValid={this.state.inputValidity.isCorrectAnswerValid}
            answerA={this.state.answerA} isAnswerAValid={this.state.inputValidity.isAnswerAValid}
            answerB={this.state.answerB} isAnswerBValid={this.state.inputValidity.isAnswerBValid}
            answerC={this.state.answerC} isAnswerCValid={this.state.inputValidity.isAnswerCValid}/>;
    }

    getErrorMessage = () => {
        return <ErrorMessage message={this.props.errorMessage}/>;
    }

    getFormFooter = () => {
        return React.cloneElement(this.FormFooter, {
            clickHandler: (e) => this.handleClickEvent(e),
            cancelButtonText: 'Clear All',
            cancelButtonID: 'clearAllButton'
        });
    }

    handleInputChange = (event) => {
        const { target: { id, value }} = event;

        if (id === 'questionType') {
            this.setState({ [id]: value, correctAnswer: '', answerA: '', answerB: '', answerC: '' });
        } else {
            this.setState({ [id]: value });
        }
    };

    handleClickEvent = (event) => {
        event.preventDefault();

        if (event.target.id === 'clearAllButton') {
            this.setState(this.initialState);
            this.props.clearErrorMessage();
        } else if (this.isAllInputValid()){
            this.props.submit(event.target.id);
        }

    }

    isAllInputValid = () => {
        let inputValidity = this.getInputValidity();
        this.setState({ inputValidity });

        for (const entry in inputValidity) {
            if (!inputValidity[entry]) {
                return false;
            }
        }

        return true;
    };

    getInputValidity = () => {
        return {
            isQuestionTypeValid: this.state.questionType.length > 0 ? true : false,
            isQuestionDescriptionValid: this.state.questionDescription.length > 9  ? true : false,
            isCorrectAnswerValid: this.state.correctAnswer.length > 0 ? true : false,
            isAnswerAValid: this.state.answerA.length > 0 ? true : false,
            isAnswerBValid: this.state.answerB.length > 0 ? true : false,
            isAnswerCValid: this.state.answerC.length > 0 ? true : false,
            isMajorTopicValid: this.state.topic.majorCategory.length > 0 ? true : false,
            isSubTopicValid: this.state.topic.subCategory.length > 0 ? true : false
        };
    };
    
}

export default IBCSQuestionForm;