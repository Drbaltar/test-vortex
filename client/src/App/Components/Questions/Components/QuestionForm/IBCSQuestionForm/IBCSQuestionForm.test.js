import React from 'react';
import { shallow } from 'enzyme';
import IBCSQuestionForm from './IBCSQuestionForm';

const baseQuestionFieldValues = {
    questionType: 'Multiple Choice',
    questionDescription: '',
    correctAnswer: '',
    answerA: '',
    answerB: '',
    answerC: ''
};

const baseQuestionFieldValidity = {
    isQuestionTypeValid: null,
    isQuestionDescriptionValid: null,
    isCorrectAnswerValid: null,
    isAnswerAValid: null,
    isAnswerBValid: null,
    isAnswerCValid: null
};

const invalidBaseQuestionFieldValidity = {
    isQuestionTypeValid: true,
    isQuestionDescriptionValid: false,
    isCorrectAnswerValid: false,
    isAnswerAValid: false,
    isAnswerBValid: false,
    isAnswerCValid: false
};

const expectedInitialState = {
    ...baseQuestionFieldValues,
    testType: {
        tactics: false,
        earlyWarning: false,
        weaponsControl: false
    },
    topic: {
        majorCategory: '',
        subCategory: ''
    },
    inputValidity: {
        ...baseQuestionFieldValidity,
        isTestTypeValid: null,
        isMajorTopicValid: null,
        isSubTopicValid: null
    },
    isTopicLoading: false,
    existingTopicCategories: null
};

const stateWithValidMultChoiceFields = {
    questionType: 'True or False',
    questionDescription: 'What is my name?',
    correctAnswer: 'Kyle',
    answerA: 'Troy',
    answerB: 'Gavin',
    answerC: 'Mark',
    testType: {
        tactics: false,
        earlyWarning: true,
        weaponsControl: false
    },
    topic: {
        majorCategory: 'Test',
        subCategory: 'Categories'
    }
};

class TestFooter extends React.Component {
    constructor(props) {
        super(props);
    }

    render() { <div />; }
}

const mockPreventDefault = jest.fn();
const mockClearErrorMessage = jest.fn();
const mockSubmit = jest.fn();

describe('IBCSQuestionForm', () => {
    const wrapper = shallow(<IBCSQuestionForm title='Add New IBCS Question' submit={mockSubmit} clearErrorMessage={mockClearErrorMessage}>
        <TestFooter />
    </IBCSQuestionForm>);

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('sets the state to the correct initial state', () => {
        expect(wrapper.state()).toEqual(expectedInitialState);
    });

    it('renders the BodyCard component', () => {
        expect(wrapper.find('BodyCard').exists()).toBe(true);
    });

    it('passes the title prop to the BodyCard component', () => {
        expect(wrapper.find('BodyCard').prop('title')).toBe('Add New IBCS Question');
    });

    it('renders the BaseQuestionFields component', () => {
        expect(wrapper.find('BaseQuestionFields').exists()).toBe(true);
    });

    it('passes the correct prop values to the BaseQuestionFields component', () => {
        expect(wrapper.find('BaseQuestionFields').props()).toMatchObject({...baseQuestionFieldValues, ...baseQuestionFieldValidity});
    });

    it('passes the input handler function to the BaseQuestionFields component', () => {
        expect(wrapper.find('BaseQuestionFields').prop('inputHandler')).toBeInstanceOf(Function);
    });

    describe('when the question type field changes', () => {
        beforeAll(() => {
            wrapper.setState({ correctAnswer: 'ECS', answerA: 'ECS', answerB: 'EPP', answerC: 'BCP' });

            const inputHandler = wrapper.find('BaseQuestionFields').prop('inputHandler');
            inputHandler({ target: { id: 'questionType', value: 'Fill-in-the-Blank'}});
        });

        it('sets the correctAnswer to blank', () => {
            expect(wrapper.state().correctAnswer).toBe('');
        });

        it('sets the alternate answers to blank', () => {
            expect(wrapper.state().answerA).toBe('');
            expect(wrapper.state().answerB).toBe('');
            expect(wrapper.state().answerC).toBe('');
        });

        afterAll(() => {
            wrapper.setState({
                questionType: baseQuestionFieldValues.questionType,
                correctAnswer: baseQuestionFieldValues.correctAnswer,
                answerA: baseQuestionFieldValues.answerA,
                answerB: baseQuestionFieldValues.answerB,
                answerC: baseQuestionFieldValues.answerC
            });
        });
    });

    it('renders the CheckBoxGroup component', () => {
        expect(wrapper.find('CheckBoxGroup').exists()).toBe(true);
    });

    it('passes the correct props to the CheckBoxGroup', () => {
        expect(wrapper.find('CheckBoxGroup').props()).toEqual({
            heading: 'Please select the tests for which this question is applicable:',
            boxGroup: [
                {
                    id: 'tactics',
                    label: 'Tactics',
                    value: false
                },
                {
                    id: 'earlyWarning',
                    label: 'Early Warning',
                    value: false
                },
                {
                    id: 'weaponsControl',
                    label: 'Weapons Control',
                    value: false
                }
            ],
            checkboxChange: expect.any(Function),
            isValid: expectedInitialState.inputValidity.isTestTypeValid,
            errorMessage: 'At least one \'Test Type\' must be selected!'
        });
    });

    describe('when each test type checkbox has been selected', () => {
        const checkboxChange = wrapper.find('CheckBoxGroup').prop('checkboxChange');
        
        it('changes the tactics variable in state', () => {
            checkboxChange({ target: { id: 'tactics', checked: true }});

            expect(wrapper.state().testType.tactics).toBe(true);
        });

        it('changes the earlyWarning variable in state', () => {
            checkboxChange({ target: { id: 'earlyWarning', checked: true }});

            expect(wrapper.state().testType.earlyWarning).toBe(true);
        });

        it('changes the weaponsControl variable in state', () => {
            checkboxChange({ target: { id: 'weaponsControl', checked: true }});

            expect(wrapper.state().testType.weaponsControl).toBe(true);
        });

        afterAll(() => {
            wrapper.setState({ testType: expectedInitialState.testType });
        });
    });

    it('passes the footer component to the BodyCard component', () => {
        expect(wrapper.find('BodyCard').childAt(1).name()).toEqual('TestFooter');
    });

    it('passes the correct props to the FormFooter component', () => {
        const expectedProps = {
            clickHandler: expect.any(Function),
            cancelButtonText: 'Clear All',
            cancelButtonID: 'clearAllButton'
        };
    
        expect(wrapper.find('TestFooter').props()).toEqual(expectedProps);
    });

    describe('when the click handler is called from the Form Footer', () => {
        describe('when the clearAllButton is clicked', () => {
            beforeAll(() => {
                const clickhandler = wrapper.find('TestFooter').prop('clickHandler');

                wrapper.setState(stateWithValidMultChoiceFields, () => clickhandler({
                    target: { id: 'clearAllButton'},
                    preventDefault: mockPreventDefault
                }));
            });

            it('prevents default behavior', () => {
                expect(mockPreventDefault).toHaveBeenCalled();
            });

            it('calls the clearErrorMessage prop', () => {
                expect(mockClearErrorMessage).toHaveBeenCalled();
            });

            it('sets the state back to the initial value', () => {
                expect(wrapper.state()).toEqual(expectedInitialState);
            });

            afterAll(() => {
                mockPreventDefault.mockClear();
                mockClearErrorMessage.mockClear();
            });
        });

        describe('when any other button is clicked and all fields are valid', () => {
            beforeAll(() => {
                wrapper.setState(stateWithValidMultChoiceFields);
                const clickhandler = wrapper.find('TestFooter').prop('clickHandler');

                clickhandler({
                    target: { id: 'submitButton'},
                    preventDefault: mockPreventDefault
                });
            });

            it('prevents default behavior', () => {
                expect(mockPreventDefault).toHaveBeenCalled();
            });

            it('calls the submit prop with the button id', () => {
                expect(mockSubmit).toHaveBeenCalledWith('submitButton');
            });

            afterAll(() => {
                mockPreventDefault.mockClear();
                mockSubmit.mockClear();
                wrapper.setState(expectedInitialState);
            });
        });

        describe('when any other button is clicked and all fields are NOT valid', () => {
            beforeAll(() => {
                const clickhandler = wrapper.find('TestFooter').prop('clickHandler');

                clickhandler({
                    target: { id: 'submitButton'},
                    preventDefault: mockPreventDefault
                });
            });

            it('prevents default behavior', () => {
                expect(mockPreventDefault).toHaveBeenCalled();
            });

            it('does not call the submit prop', () => {
                expect(mockSubmit).not.toHaveBeenCalled();
            });

            it('passes the correct prop to the BaseQuestionFields component', () => {
                expect(wrapper.find('BaseQuestionFields').props()).toMatchObject(invalidBaseQuestionFieldValidity);
            });

            it('passes the correct prop to the CheckBoxGroup', () => {
                expect(wrapper.find('CheckBoxGroup').prop('isValid')).toBe(false);
            });

            afterAll(() => {
                mockSubmit.mockClear();
                mockPreventDefault.mockClear();
            });
        });

        describe('when an error message is passed in, it is displayed', () => {
            const testErrorMessage = 'There was an error when submitting';

            beforeAll(() => {
                wrapper.setProps({ errorMessage: testErrorMessage });
            });

            it('renders an ErrorMessage component', () => {
                expect(wrapper.find('ErrorMessage').exists()).toBe(true);
            });

            it('passes the error message to the ErrorMessage component', () => {
                expect(wrapper.find('ErrorMessage').prop('message')).toBe(testErrorMessage);
            }); 
        });
    });
});