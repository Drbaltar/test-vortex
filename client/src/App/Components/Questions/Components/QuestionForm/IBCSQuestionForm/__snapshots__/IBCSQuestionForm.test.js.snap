// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IBCSQuestionForm renders correctly 1`] = `
<BodyCard
  title="Add New IBCS Question"
>
  <div
    id="issue-form-body"
  >
    <BaseQuestionFields
      answerA=""
      answerB=""
      answerC=""
      correctAnswer=""
      inputHandler={[Function]}
      isAnswerAValid={null}
      isAnswerBValid={null}
      isAnswerCValid={null}
      isCorrectAnswerValid={null}
      isQuestionDescriptionValid={null}
      isQuestionTypeValid={null}
      questionDescription=""
      questionType="Multiple Choice"
    />
  </div>
  <TestFooter
    cancelButtonID="clearAllButton"
    cancelButtonText="Clear All"
    clickHandler={[Function]}
  />
</BodyCard>
`;
